src/helpers/calculationStepFunctions/numericCalculationFunctions.ts

```diff
@@ -0,0 +1,54 @@
+import { create, all } from 'mathjs'
```

Comment:
oh neat find with mathjs - I like it better than using eval.

So actually...all this string parsing logic should be unnecessary?  Because mathjs supports passing variables?  Like here's one of the examples:

```
let scope = {
    a: 3,
    b: 4
}
math.evaluate('a * b', scope)           // 12
math.evaluate('c = 2.3 + 4.5', scope)   // 6.8
scope.c                                 // 6.8
```

So you could imagine setting up your scope as:
```
let scope = {
    '[Q1]': 3,
    '[Q2]': 4,
    '[F1.Q1]': 5
}
```
from your survey answers.  For form variables (where step id/result id are different) one could just use a dot separator like [F1.Q1].

Or you can set up the scope to call a function to get the answer it looks like: https://mathjs.org/examples/advanced/custom_scope_objects.js.html

The main question/issue is whether mathjs is kind/permissive enough to allow numbers in it's symbols; seems like there is some flexibility there but not sure how much.

Reply:
Cool. I've been playing with this and have a few thoughts on feasibility.

1. the `.` is the object access operator. So this seems to prevent us from being able to do calculations with stepIds/resultIds that have a `.` in the name (ie. `Form1.Q1.2`). I think we can probably add a little validation message that says when you have calculation step, the calculation string cannot reference a stepIdentifier/resultIdentifier combo that has a `.` in it. 
^How does that sound?

2. When assembling the scope dictionary from the survey answers, does this rule work? if stepIdentifier only appears once in the survey answer list, then use just `{stepIdentifier: value}`, otherwise if it appears more than once use `stepIdentifier: {resultIdentifier1: value, resultIdentifier2: value}`

Reply:
@Suvarna-Thejas-CE maybe dig into REDCap a bit here; what kind of validation / restrictions do they have. 

Even might be worth it if we follow their syntax for the vars; not sure if math.js will allow []  though, e.g. ```[clab_lipidldl_u2] * [clab_lipidldlab]```

For 2...after thinking about it...I think we should just key it by result identifier.  It makes it so much simpler...and I would add validation so that your result identifiers have to be unique (e.g. you can't have Q1 on FormStep1 and Q1 on FormStep2) if you have a numeric calculation step.  

Thoughts?

Reply:
On my preliminary testing, the `[]` do work in mathJS :) so that should work great. Will look into REDCap a bit here. 

I like **not** having restrictions on the step/resultIdentifiers. The '_' is a cool idea to me and feels like a great one from a consumer perspective.

For me, the usability concern for using only resultIdentifiers is that the question step UI doesn't mention "resultIdentifier". Could potentially be fixed with documentation, but would choose not to force customers to rely on reading user guide articles.

Reply:
Ah hmm I wasn't suggesting we do anything with underscores actually...was not a separator I was proposing, sorry.  Those were just some redcap vars I copy/pasted.  

Basically just thinking that we have [resultIdentifier] be the syntax; and we enforce unique result identifiers when using this step because in practice they are anyways, and it will be a pain to construct a step identifier / result identifier pair (e.g. [formstepidentifier_resultidentifier] or whatever you were thinking)

In terms of usability, I think that might just be how we describe it.  It is either the step identifier OR the form item identifier if it is a form step basically; so we can avoid the use of "resultIdentifier" if we want (or we can clarify in the survey builder as well).  But it again feels easier to me than the user having to write out [formstepidentifier_resultidentifier] if they can just always do [resultidentifier].  

We don't necessarily have to enforce it via a validation rule.  We can just have [resultidentifier] be the way this works and if they happen to have multiple overlapping resultidentifiers, it just takes the latest survey answer with that resultidentifier.

Reply:
For what it's worth; using only the result identifier is what I did on my version of this since I knew all my questions were form questions, and the result identifiers were unique.  What stinks about that is that there's not a good way to find the right question that the result identifier is on (there's lots of repeated looping).  I debated, in my version, making a dictionary to do lookups so that I only did the looping once, but didn't take it that far.

That said, I'm also somewhat in favor of only the result identifier being in the eventual UI because, just as an example, the survey I was working on required doing a sum total of 91 variables for its final calculation.  Listing 91 variable names is pretty gnarly in a text input; it's even worse if they're longer because they require the question identifier as well.

Reply:
Cool - yeah. Y'all are convincing me that there is value to using ResultIdentifier. The functional name seems to be "QuestionIdentifier" which, like you said Chris, is defined as the StepIdentifier for Question Steps or FormItemIdentifier for Form Steps. I think this is workable.

It does bring up some questions about taxonomy for me (that I want to document, and maybe get reactions too, but is a bit of scope creep for this PR specifically). 

In the case, where I have repeated questions (i.e, For all the people in your family, specify the height and weight). I would do something that looks like:
```
Form Step: Person1
 -- Form Item: Height
 -- Form Item: Weight
Form Step: Person2
 -- Form Item: Height
 -- Form Item: Weight
Form Step: Person3
 -- Form Item: Height
 -- Form Item: Weight
```
Currently, it's not obvious that there would be a relationship between a form item identifier and a question step identifier. 

I think it makes sense to make that relationship explicit, just through naming changes. 

- Renaming Form Item to Question or Question Item 
- Potentially renaming Form Step to multi-Question step

It would change the taxonomy from:
```
- Question Step
   -- Question
- Form Step
  -- Form Item
     --- Question
  -- Form Item
     --- Question
```

To this model that makes the relationship between multi-question and question steps clear. Functionally, I think it also describes better what a form step is, just a collection of questions. 
 ```
- Question Step
   -- Question Identifier
- Multi-Question Step Identifier
   -- Question Identifier
   -- Question Identifier
```

Reply:
Well...we cannot go renaming things like crazy unfortunately...have to consider the downstream implications given that creates inconsistencies with exported data and the survey answer API.  Have to think of the scope of the change we are making here

Reply:
Maybe we can give this another quick review after the latest push